### PR TITLE
Add workflowengine

### PR DESCRIPTION
**Pull request by MorrisJobke**
cc @nickvergessen

### CODE REVIEW

**Comment by MorrisJobke** [comment]
https://github.com/nextcloud/files_accesscontrol <- this app then shows the UI for the workflow engine with regard to the access control rules.

File: apps/workflowengine/appinfo/routes.php [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+return [
+	'routes' => [
+		['name' => 'flowOperations#getChecks', 'url' => '/checks', 'verb' => 'GET'], // TODO rm and do via js?
```
Comment (nickvergessen) [inline apps/workflowengine/appinfo/routes.php]: todo?

File: apps/workflowengine/js/admin.js [hunk 2]
```diff
@@ -0,0 +1,367 @@
...
+				for (var i = 0; i < event.target.classList.length; i++) {
+					var className = event.target.classList[i];
+					if (className.substr(0, 'check-'.length) === 'check-') {
+						key = className.substr('check-'.length);
+						break;
+					}
+				}
+
+				if (key === null) {
+					console.warn('checkChanged triggered but element doesn\'t have any "check-" class');
```
Comment (nickvergessen) [inline apps/workflowengine/js/admin.js]: Delete all console?

File: apps/workflowengine/templates/admin.php [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+	</script>
+
+	<script type="text/template" id="operation-template">
+		<div class="operation{{#if hasChanged}} modified{{/if}}">
+			<input type="text" class="operation-name" value="{{operation.name}}">
+			{{! delete only makes sense if the operation is already saved }}
+			{{#if operation.id}}
+			<span class="button-delete pull-right icon-delete"></span>
+			{{/if}}
+			<span class="pull-right info">{{operation.class}} - ID: {{operation.id}} - operation: {{operation.operation}}</span>
```
Comment (nickvergessen) [inline apps/workflowengine/templates/admin.php]: This looks debugish.

I assume operations may need to be of a input for other things, like autotagging (so you can select2 the tags?)

File: apps/workflowengine/templates/admin.php [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+				<span class="pull-right">Saving ...</span>
+			{{else}}{{#if message}}
+				<span class="msg pull-right {{#if errorMessage}}error{{else}}success{{/if}}">
+					{{message}}{{#if errorMessage}} {{errorMessage}}{{/if}}
+				</span>
+			{{/if}}{{/if}}
+		</div>
+	</script>
+
+	<div class="rules"><span class="icon-loading-small"></span> Loading ...</div>
```
Comment (nickvergessen) [inline apps/workflowengine/templates/admin.php]: Missing L10N everywhere in here

File: core/shipped.json [hunk 5]
```diff
@@ -27,11 +27,13 @@
...
-    "user_saml"
+    "user_saml",
+    "workflowengine"
   ],
   "alwaysEnabled": [
     "files",
     "dav",
-    "federatedfilesharing"
+    "federatedfilesharing",
+    "workflowengine"
```
Comment (nickvergessen) [inline core/shipped.json]: to discuss I guess, but good enough for now

**Comment by nickvergessen** [comment]
Tests fail Oo 
> PHP Fatal error:  Call to a member function getUpdater() on a non-object in /drone/src/github.com/nextcloud/server/lib/private/Files/View.php on line 695

Integration tests is missing the app on the provisioning tests

File: apps/workflowengine/templates/admin.php (hunk 3 above)
Comment (MorrisJobke) [inline apps/workflowengine/templates/admin.php]: Yes and no: I still don't know to what the operation attribute can be set. Then I could add this as input.

File: apps/workflowengine/templates/admin.php (hunk 4 above)
Comment (MorrisJobke) [inline apps/workflowengine/templates/admin.php]: RIght.

**Comment by nickvergessen** [comment]
Created issues for everything, lets just fix the tests here, thats the most important.

**Comment by MorrisJobke** [comment]
> Created issues for everything, lets just fix the tests here, thats the most important.

If I comment out https://github.com/nextcloud/server/blob/55fc6536d33bd044a72437ac61d3c5ade09111cb/tests/lib/AppTest.php#L433-L466 all works fine 🙈

File: apps/workflowengine/js/admin.js (hunk 2 above)
Comment (MorrisJobke) [inline apps/workflowengine/js/admin.js]: I would like to keep them. Makes it then easier to debug once a problem is noticed. But I'm fine with removing.

**Comment by nickvergessen** [comment]
:+1:

File: apps/workflowengine/appinfo/routes.php (hunk 1 above)
Comment (MorrisJobke) [inline apps/workflowengine/appinfo/routes.php]: How should this be done via JS? -> let's move this to another issue/PR

Comment (nickvergessen) [inline apps/workflowengine/appinfo/routes.php]: Already made a PR in #582

**Comment by rullzer** [comment]
Fine by me if the mentioned stuff gets tackled.
LGTM

**Pull request merged**